usr/src/Xapps/xview3/clients/cmdtool/cmdtool.man
.\" @(#)cmdtool.1 1.27 90/06/21 SMI; 
.\" @(#)cmdtool.1 1.25 90/02/26 SMI;
.\" Updated 6/8/90 
.TH cmdtool 1 "11 June 1990"
.SH NAME
.B cmdtool 
\- run a shell (or other program) in an OpenWindows enhanced terminal window
.SH SYNOPSIS
.B cmdtool
[
.B \-C
]
[
.B \-M
.I bytes
]
[
.B \-P
.I count
]
[
.B \-B
.I boldstyle
]
[
.B \-I
.I command
]
[
.I generic-tool-arguments
]
[
.I program
[ 
.I program-arguments
]
]
.SH AVAILABILITY
This command is available with the
OpenWindows user environment.
For information about installing OpenWindows, refer to the
.I OpenWindows Installation and Start-Up Guide.
.SH DESCRIPTION
.IX cmdtool "" "\fLcmdtool\fR \(em shell or program with XView text facility"
.IX "shell window" "\fLcmdtool\fR"
.LP
.B cmdtool
is the standard
.I OpenWindows
support facility for shells and other programs.
When invoked,
.B cmdtool
runs a program (usually a shell) in a text-based command window.
Characters typed on the keyboard are inserted at the caret.  If the
program is a shell, that shell accepts and runs commands in the
usual manner.  
.B cmdtool
also supports programs that perform cursor motions directly, such
as
.BR vi (1).
.LP
The text of the current command line can be edited using normal
.BR textedit (1)
functions.
The command window displays a log of the
session, which can be scrolled through using the scrollbar
(unless the escape command mode is in effect).
This log can be edited, and saved by choosing the
.RB ` "Store as New File" '
item in the text facility's pop-up menu.
.SH OPTIONS
.TP 15
.B \-C
Console 
.BR cmdtool .
Redirect system console output to this shelltool.
Display console messages in this
.BR cmdtool ,
which might otherwise appear in unexpected places on the 
workstation screen.  Since a
.B cmdtool
window can be scrolled, console error messages can be recorded for
later examination.
.TP
.BI \-M " bytes"
Set the log to wrap-around after the indicated
number of
.IR bytes .
.TP
.BI \-P " count"
Checkpoint the log after every set of
.I count
editing operations.
.TP
.BI \-B " boldstyle"
Set the style for displaying bold text to
.IR boldstyle .
.I boldstyle
can be a string specifying one of the choices for the
.B term.boldstyle
default, see
.BR "Defaults Options" ,
below, or it may be a numerical value for one of those choices,
from 0 to 8, corresponding to the placement of the choice in the list.
.TP
.BI \-I " command"
Pass
.I command
to the shell. 
.SM SPACE
characters within the command must be escaped.
.TP
.I generic-tool-arguments
.B cmdtool
accepts the generic tool arguments
listed in
.BR xview (1).
.HP
.I program
[
.I program-arguments
]
.br
If a
.I program
argument is present,
.B cmdtool
runs it and passes any remaining arguments to that
.IR program .
If no
.I program
is given,
.B cmdtool
runs the program indicated by the
.B \s-1SHELL\s0
environment variable, or
.B /bin/sh
by default.
.SH USAGE
.SS .Xdefaults File Options
.LP
You can specify a number of defaults using the options
in the 
.B .Xdefaults
file that effect the behavior of 
.BR  cmdtool .
The ones of interest are those that begin with
.BR text , term ,
or 
.BR keyboard .
See
.BR xview (1)
for more detailed information.
.SS cmdtool Windows
.LP
The window created by 
.B cmdtool
is based on the text facility that is described in the 
.B textedit
man page.
The user is given a prompt at which to type commands and pop-up
menus from which to select command options.
.LP
.B cmdtool
windows support cursor motions, using an
.B /etc/termcap
entry called
.BR sun-cmd .
Command windows automatically set the
.B \s-1TERM\s0
environment variable to
.BR sun-cmd .
So, if you
.BR rlogin (1C)
to a machine that does not have an entry for
.B sun-cmd
in its
.B /etc/termcap
file, the error message
.RB ` "Type sun-cmd unknown" '
r
.RB ` "set \s-1TERM\s0=sun" '.
Programs written using the
.BR curses (3X)
or 
.BR curses (3V)
library packages will work in a command window, but programs
hard-coded for
.BR sun -type
terminals may not. 
When supporting a program that performs
cursor motions, the command window automatically takes on the
characteristics of a tty window (as with
.BR shelltool (1)).
When that program terminates or sleeps, the full command window
functionality is restored.
.LP
.B cmdtool
supports programs that use
.SM CBREAK
and
.SM NO ECHO
terminal modes. 
This support is normally invisible to the user.
However, programs that use
.SM RAW
mode, such as
.BR rlogin (1C)
and
.BR script (1),
inhibit command-line editing with the mouse.
In this case, however, tty-style
.SM ERASE,
word-kill and line-kill characters can still be used to edit the
current command line.
.SS The cmdtool Menu
.LP
The
.B cmdtool
window menu is called the
.B Term Pane
menu and contains the following options and their submenus:
.TP 10
.PD 0
.B History 
Creates a a list of commands used during the
cmdtool session.
.RS
.TP 10
.B Mode
.RS
.TP 10
.B Editable
You can edit the contents of the window.
.TP
.B Read Only
You can only read from the window.
.RE
.TP
.B Store Log as new file
Create a new file that contains the contents of the log.
.TP
.B Clear log
Clears all entries from the log.
.RE
.TP
.B Edit
Provides a set of editing functions for this window.
.RS
.TP 10
.B Again
Repeats the last action.
.TP
.B Undo
.RS
.TP 10
.B Undo Last Edit
Undoes the last edit made in
.B cmdtool .
.TP
.B Undo All Edits
Undoes all edits made during this session of
.B cmdtool .
.RE
.TP
.B Copy
Makes a copy of the selected text and places 
it on the clipboard.
.TP
.B Paste
Pastes a copy of the text stored on the 
clipboard at the cursor location.
.TP
.B Cut
Deletes the selected text.RE
.TP
.sp
Provides a set of find and replace functions.
.RS
.TP 10
.B Find and Replace
Brings up a pop-up menu containing text fields 
and command buttons that allow you to search 
forward and backward through the file being 
edited for specific text strings. 
Allows you to specify options for the 
replacement of text.
.TP
.B Find Selection
.RS
.TP 10
.B Forward
Searches forward to find a selected text string.
.TP
.B Backward
Searches backward to find a selected text string.
.RE
.TP
.B Find Marked Text
Brings up a pop-up menu that allows you to 
find text that is included between specified 
bracket pairs. 
Also allows you to insert or remove bracket 
pairs around selected text.
.TP
.B Replace |>field<| >
Allows you to replace selected text forward 
and backward throughout the file.
.RE
.RE
.TP 10
.B Extras
A user-definable pull-right menu controlled by the
.B /usr/lib/.text_extras_menu
file.
This can be overridden in two ways:
.br
1) Change the value of the .Xdefaults parameter 
.B text.extrasMenuFilename 
to the correct file path.
.br
2) Set the environment variable 
.B \s-1EXTRASMENU\s0
to the file desired.
.br
Note that option 1 overrides option 2 if both are
used.
For more information see the
.I "DeskSet Environment Reference Guide" .
.TP
.B File Editor
.RS
.TP 10
.B Enable
Allows you to edit files from within
.B cmdtool.
.TP
.B Disable
Turns off the ability to edit files from
within 
.B cmdtool.
.RE
.TP
.B Scrolling
.RS
.TP 10
.B Enables Scrolling
Enables scrolling within
.B cmdtool.
.TP
.B Disable Scrolling
Turns off the ability to scroll within
.B cmdtool.
Once scrolling in 
.B cmdtool is disabled, it's functionality is
identical to
.B shelltool
and a more restricted menu appears. Selecting the
.B Enable Scrolling
option from the restricted menu restores the full menu
and functionality of
.B cmdtool.
.R.RE
.SS User Defined Keyboard Remapping
The file 
.B ~/.textswrc
specifies filter programs that are assigned to
(available) function keys.
These filters are applied to the contents of the
primary selection.
Their output is entered at the caret.
.SS Accelerators
.LP
Text facility accelerators that are especially useful in
command windows are described here.  See 
.br
.BR textedit (1)
for more information.
.TP 18
.SM CTRL-RETURN
Position the caret at the bottom, and scroll it into view
as determined by 
.BR Text.LowerContext .
.TP 
.PD 0
\s-1CAPS\s0-lock
.TP
Toggle between all-upper-case keyboard input, and mixed-case.
.PD
.SH FILES
.PD 0
.TP 20
.BI /tmp/tty.txt. pid
log file
.TP
.B ~/.textswrc
.TP
.B ~/.ttyswrc
.TP
.B usr/lib/.text_extras_menu
.TP
.B $HOME/.TextExtraMenu
.TP
.B /etc/termcap
.TP
.B /usr/bin/sh
.PD
.SH "SEE ALSO"
.BR rlogin (1C) ,
.BR script (1),
.BR shelltool (1),
.BR textedit (1) , 
.BR vi (1) ,
.BR xview (1) ,
.BR curses (3V) ,
.BR curses (3X)
.LP
.I "OpenWindows User's Guide" 
.SH BUGS
.LP
Typing ahead while
.B cmdtool 
changes between its scrollable and cursor motion modes
will sometimes freeze
.B cmdtool .
.LP
Full terminal emulation is not complete.
Some manifestations of this deficiency are:
.TP 3
\(bu
File completion in the C shell does not work.
.TP
\(bu
Enhanced display of text is not supported.


usr/src/Xapps/xview3/clients/cmdtool/cmdtool.c
#ifndef lint
#ifdef sccs
static  char sccsid[] = "@(#)cmdtool.c 15.57 91/09/14";
#endif
#endif

/*
 * Copyright (c) 1985, 1987 by Sun Microsystems, Inc.
 */

/*
 *  cmd/shelltool - run a process in a tty subwindow
 */

#include <stdio.h>

#ifndef SVR4
#include <strings.h>
#else
#include <string.h>
#endif SVR4

#include <sys/types.h>
#include <xview/attr.h>
#include <xview/defaults.h>
#include <xview/xview.h>
#include <xview/scrollbar.h>
#include <xview/tty.h>
#include <xview/termsw.h>
#include <xview/textsw.h>

#define	TEXTDOMAIN	"cmdtool"
#define	MSG(msg)	dgettext(TEXTDOMAIN, msg)

extern	char *getenv();
extern char *dgettext();

static unsigned short tty_image[258] = {
#include <images/terminal.icon>
};

static unsigned short tty_mask_image[258] = {
#include <images/terminal_mask.icon>
};

static unsigned short cmd_image[258] = {
#include <images/cmdtool.icon>
};

static unsigned short cmd_mask_image[258] = {
#include <images/cmdtool_mask.icon>
};

static unsigned short console_image[258] = {
#include <images/console.icon>
};

static unsigned short console_mask_image[258] = {
#include <images/console_mask.icon>
};


print_usage(am_cmdtool, toolname)
	int	 am_cmdtool;
	char	*toolname;
{
	char	*mode_spec = (am_cmdtool)
				? MSG("-P frequency") :
				MSG("-B boldstyle");
	
	(void)fprintf(stderr,
		MSG("syntax: %s [-C] [-I initial_cmd] [%s] [program [args]]\n"),
		toolname, mode_spec);
	if (!am_cmdtool) {
	    (void)fprintf(stderr,
		MSG("\
-B	set bold style for this instance of %s\n\
	where 'boldstyle' is a number from 1 to 8\n"), toolname);
	}
	(void)fprintf(stderr,
		MSG("-C	redirect console output to this instance of %s\n"),
		toolname);
	(void)fprintf(stderr,
		MSG("-I	'initial_cmd' is first command 'program' executes\n"));
	(void)fprintf(stderr,
		MSG("-L	run the .login file when starting the shell subprocess\n"));
	if (am_cmdtool) {
	    (void)fprintf(stderr,
		MSG("\
-P	checkpoint frequency for this %s, where 'frequency' is number\n\
	of edits between checkpoints; a value of 0 means no checkpointing.\n"),
		toolname);
	}
}

main(argc,argv)
	int argc;
	char **argv;
{
	int	am_cmdtool;
	Frame	base_frame;
	Tty	ttysw;
	Icon	tool_icon;
	char	*tool_name = argv[0];
	char	*shell_label;
	char	*cmd_label;
	char	*console_label;
	char	frame_label[150];
	char	icon_label[30];
        char    *tmp_label1, *tmp_label2;
	int	become_console = 0;
	int	run_login	= 0;
	char	*bold_name = 0;
	char	*sh_argv[2];
	char	*init_cmd = 0;
	int	len;
	char	*filename = (char *)strrchr(argv[0], '/');
	int	checkpoint = 0;
	int	edit_log_wraps_at = TEXTSW_INFINITY;
	int	tty_pid = 0;
	char	err_msg[50];
	char    *cmdline[100];
	int     cmdline_count = 0;
	char	cmdline2[50];
	Server_image  cmd_pixmap, cmd_mask_pixmap;

	
#ifdef GPROF
	if (argc > 1 && strcmp(argv[argc-1], "-gprof") == 0) {
	    moncontrol(1);
	    /* Pull the -gprof out of argc/v */
	    argc--;
	    argv[argc] = (char *)0;
	} else {
	    moncontrol(0);
	}
#endif
	
#ifdef DEBUG
	malloc_debug(0);   
#endif	
	
	/* This is required to initialize correctly */	
	xv_init(XV_INIT_ARGC_PTR_ARGV, &argc, argv, 
		XV_USE_LOCALE, TRUE, 0);

	shell_label = MSG("shelltool");
	cmd_label = MSG("cmdtool");
	console_label = MSG(" (CONSOLE) - ");

	if (filename)
		filename++;
	else
		filename = argv[0];
	am_cmdtool = (strcmp(filename, "cmdtool") == 0);

	/*
	 *  Send the icon attr before argc, argv to give
	 *  commandline argument a chance to override.
	 *  A waste of space & time if commandline argument
	 *  is present.
	 */
	icon_label[0] = 0177;	/* del, highly unlikely as cmd arg */
	icon_label[1] = '\0';

        cmd_pixmap = (Server_image)xv_create(NULL, SERVER_IMAGE,
                   	                     XV_WIDTH,  64,
		                             XV_HEIGHT, 64,
		                             SERVER_IMAGE_BITS,      
		                             become_console ? console_image : 
                                             am_cmdtool ? cmd_image : tty_image,
		                             NULL);

        cmd_mask_pixmap = (Server_image)xv_create(NULL, SERVER_IMAGE,
		                                 XV_WIDTH,  64,
		                                 XV_HEIGHT, 64,
		                                 SERVER_IMAGE_BITS,
		                                 become_console ? console_mask_image : 
                                                 am_cmdtool ? cmd_mask_image : 
                                                              tty_mask_image,
		                                 NULL);

	tool_icon = (Icon)xv_create(NULL, ICON,
            WIN_RETAINED, TRUE,				    
            ICON_IMAGE, cmd_pixmap,
            ICON_MASK_IMAGE, cmd_mask_pixmap,
            ICON_LABEL, icon_label,
            ICON_TRANSPARENT,TRUE,
            NULL);

	base_frame = xv_create((Xv_window)NULL, FRAME,
			FRAME_ICON, tool_icon,
		        HELP_STRING_FILENAME,	"manpage_synopsis_help_index",
			0);
			
	if (base_frame == NULL) {
	    fprintf(stderr, MSG("Cannot create base frame.  Process aborted.\n"));
	    exit(1);
	}		

	/* Get ttysw related args */
	sh_argv[0] = NULL;
	sh_argv[1] = NULL;
	argv++;
	argc--;
	if (am_cmdtool) {
	    checkpoint =
		defaults_get_integer_check("cmdtool.checkpointFrequency",
			"Term.CheckpointFrequency", 0, 0, (int)TEXTSW_INFINITY);
	    edit_log_wraps_at =
		defaults_get_integer_check("cmdtool.maxLogFileSize",
			"Term.MaxLogFileSize",
			(int)TEXTSW_INFINITY, 0, (int)TEXTSW_INFINITY);
	}
	while (argc > 0 && **argv == '-') {
		switch (argv[0][1]) {
		case 'C':
			become_console = 1;
			cmdline[cmdline_count++] = "-C";
			break;
 		case 'h':
 		case '-':	
		case 'H':
		case '?':
			print_usage(am_cmdtool, tool_name);
			(void)xv_usage(tool_name);
			exit(1);
		case 'B':
			if (argc > 1) {
				argv++;
				argc--;
				bold_name = *argv;
			        cmdline[cmdline_count++] = "-B";
			        cmdline[cmdline_count++] = bold_name;
			}
			break;
		case 'I':
			if (argc > 1) {
				argv++;
				argc--;
				init_cmd = *argv;
			}

			break;
		case 'P':
			checkpoint = atoi(argv[1]);
			cmdline[cmdline_count++] = "-P";
			cmdline[cmdline_count++] = argv[1];
			argc--, argv++;
			break;
		case 'M':
			edit_log_wraps_at = atoi(argv[1]);
			cmdline[cmdline_count++] = "-M";
			cmdline[cmdline_count++] = argv[1];
			argc--, argv++;
			break;
		case 'L':	/* jcb 5/10/90 runs .login on startup */
			run_login++;
			cmdline[cmdline_count++] = "-L";
			break;
		default:
			;
		}
		argv++;
		argc--;
	}

	if ((cmdline_count > 0) || init_cmd) {
	    if (init_cmd) {
		cmdline[cmdline_count++] = "-I";
		cmdline[cmdline_count++] = init_cmd;
	    }
	    xv_set(base_frame, FRAME_WM_COMMAND_ARGC_ARGV, 
			cmdline_count, cmdline, NULL);
	}

	if (argc == 0) {
		argv = sh_argv;
		if ((argv[0] = getenv("SHELL")) == NULL)
			argv[0] = "/bin/sh";
	}

	/* if the user wants to run .login format shell, prefix '-' to name */
	if( run_login && argv != NULL ) {	/* jcb 5/10/90 */
		strcpy( cmdline2, "-"  );
		strcat( cmdline2, argv[0] );
		argv[0]	= cmdline2;
	}

	/* If FRAME_LABEL wasn't set by cmdline argument, set it */
	if ((tmp_label1 = (char *)xv_get(base_frame, FRAME_LABEL)) == NULL) {
		(void)strncpy(frame_label,
		  am_cmdtool ? cmd_label : shell_label, sizeof(frame_label));
		if (become_console) {
			(void)strncat(frame_label, console_label,
				sizeof(frame_label));
		} else {
			(void)strncat(frame_label, " - ", sizeof(frame_label));
		}
		(void)strncat(frame_label, *argv, sizeof(frame_label));
		(void)xv_set(base_frame, FRAME_LABEL, frame_label, 0);
	}
	tool_icon = (Icon)xv_get(base_frame, FRAME_ICON);
	if (((tmp_label2 = (char *) xv_get(tool_icon, ICON_LABEL)) == NULL) ||
	    *tmp_label2 == 0177) {
	    if (tmp_label1) {
		(void)strncpy(icon_label, tmp_label1, sizeof(icon_label));
            } else if (become_console) {
		(void)strncpy(icon_label, MSG("Console"), sizeof(icon_label));
	    } else {
		(void)strncpy(icon_label, *argv, sizeof(icon_label));
	    }
	    (void)xv_set(tool_icon, ICON_LABEL, icon_label, 0);
	    xv_set(tool_icon, ICON_LABEL, icon_label, NULL);
	}
	

	ttysw = xv_create(base_frame, TERMSW,   WIN_IS_CLIENT_PANE,
		  TTY_ARGV,			argv,
		  TTY_QUIT_ON_CHILD_DEATH,	TRUE,
		  TTY_CONSOLE,			become_console,
		  0);

	if (!(defaults_exists("window.width", "Window.Width") ||
	      defaults_exists("window.height", "window.height") ||
	      defaults_exists("window.geometry", "Window.Geometry"))) {
	    int cols, rows;
	    
	    cols = defaults_get_integer_check("window.columns", "Window.Columns", 
					      80, 1, 999);
	    rows = defaults_get_integer_check("window.rows", "Window.Rows",
					      35, 1, 999);
	    
	    xv_set(ttysw,
		   WIN_COLUMNS, cols,
		   WIN_ROWS, rows,
		   0);
	    window_fit(base_frame);
	}

        if (!am_cmdtool)
		xv_set(ttysw, TERMSW_MODE, TTYSW_MODE_TYPE, 0);
	if (bold_name) {
		(void)xv_set(ttysw, TTY_BOLDSTYLE_NAME, bold_name, 0);
	}
	if (am_cmdtool) {
	    (void) xv_set(ttysw,
			  TEXTSW_CHECKPOINT_FREQUENCY, checkpoint,
			  TEXTSW_WRAPAROUND_SIZE, edit_log_wraps_at,
			  0);
	}
	
	tty_pid = (int)xv_get(ttysw, TTY_PID);
#ifdef DEBUG
	(void)fprintf(stderr, "child pid = %d\n", tty_pid);
#endif DEBUG
	if (tty_pid == -1) {
	    strcpy(err_msg, (am_cmdtool) ? MSG("Command") : MSG("Shell"));
	    strcat(err_msg, MSG(" Tool: Out of swap space.  Cannot continue.\n"));
	    (void) ttysw_output(ttysw, err_msg, strlen(err_msg));        
	} else if (init_cmd && ((len = strlen(init_cmd)) > 0)) {
	    if (init_cmd[len-1] != '\n') {
		init_cmd[len] = '\n';
		len++;
	    }
	    (void)ttysw_input(ttysw, init_cmd, len);
	}

	xv_main_loop(base_frame);
	
	exit(0);
}

